coroutines/src/commonMain/kotlin/SynchronizedMap.kt

```diff
@@ -0,0 +1,115 @@
+package com.juul.tuulbox.coroutines
+
+import kotlinx.atomicfu.locks.ReentrantLock
+import kotlinx.atomicfu.locks.reentrantLock
+import kotlinx.atomicfu.locks.withLock
+
+/** A [MutableMap] where all reads and writes are protected by a [ReentrantLock]. */
+public class SynchronizedMap<K, V>(
+    private val inner: MutableMap<K, V>,
+) : MutableMap<K, V> {
+
+    private inner class Entry(
+        val entry: MutableMap.MutableEntry<K, V>,
+    ) : MutableMap.MutableEntry<K, V> {
+        override val key: K get() = entry.key
+        override val value: V get() = entry.value
+
+        override fun setValue(newValue: V): V =
+            lock.withLock { entry.setValue(newValue) }
+    }
+
+    @PublishedApi
+    internal val lock: ReentrantLock = reentrantLock()
+
+    override val size: Int
+        get() = lock.withLock { inner.size }
+
+    override fun containsKey(key: K): Boolean =
+        lock.withLock { inner.containsKey(key) }
+
+    override fun containsValue(value: V): Boolean =
+        lock.withLock { inner.containsValue(value) }
+
+    override fun get(key: K): V? =
+        lock.withLock { inner[key] }
+
+    override fun isEmpty(): Boolean =
+        lock.withLock { inner.isEmpty() }
+
+    override val entries: MutableSet<MutableMap.MutableEntry<K, V>>
+        get() = lock.withLock {
+            inner.entries.asSequence()
+                .map { entry -> Entry(entry) as MutableMap.MutableEntry<K, V> }
```

Comment:
This cast gets highlighted as unnecessary, but also this won't compile without the cast. 🤷